csaf-import/src/main/kotlin/io/github/csaf/sbom/retrieval/RetrievedProvider.kt

```diff
@@ -49,60 +52,167 @@ class RetrievedProvider(val json: Provider) {
             }
 
     /**
-     * Validates this [RetrievedProvider] according to the CSAF standard.
+     * This function fetches all directory indices referenced by this provider.
      *
-     * @param validationContext The validation context used for validation.
+     * @param loader The instance of [CsafLoader] used for fetching of online resources.
+     * @param channelCapacity The capacity of the channels used to buffer parallel fetches. Defaults
+     *   to [DEFAULT_CHANNEL_CAPACITY].
+     * @return The fetched [Result]s, representing index contents or fetch errors.
      */
-    fun validate(validationContext: ValidationContext) {
-        role.checkRole(validationContext).let { vr ->
-            if (vr is ValidationFailed) {
-                throw vr.toException()
+    @OptIn(ExperimentalCoroutinesApi::class)
+    fun fetchDocumentIndices(
+        loader: CsafLoader = lazyLoader,
+        channelCapacity: Int = DEFAULT_CHANNEL_CAPACITY
+    ): ReceiveChannel<Pair<String, Result<String>>> {
+        @Suppress("SimpleRedundantLet")
+        val directoryUrls =
+            (json.distributions ?: emptySet()).mapNotNull { distribution ->
+                distribution.directory_url?.let { it.toString().trimEnd('/') }
+            }
+        // This channel collects up to `channelCapacity` directory indices concurrently.
+        val indexChannel =
+            ioScope.produce(capacity = channelCapacity) {
+                for (directoryUrl in directoryUrls) {
+                    send(directoryUrl to async { loader.fetchText("$directoryUrl/index.txt") })
+                }
+            }
+        // This terminal channel is a simple "rendezvous channel" for awaiting the Results.
+        return ioScope.produce {
+            for ((directoryUrl, indexDeferred) in indexChannel) {
+                send(directoryUrl to indexDeferred.await())
             }
         }
     }
 
-    /** This function fetches all CSAF documents that are listed by this provider. */
-    suspend fun fetchDocuments(loader: CsafLoader = lazyLoader): List<Result<RetrievedDocument>> {
-        @Suppress("SimpleRedundantLet")
-        return (json.distributions ?: emptySet())
-            .mapNotNull { distribution ->
-                distribution.directory_url?.let { it.toString().trimEnd('/') }
+    /**
+     * This function sums up the expected number of [RetrievedDocument]s that will be fetched from
+     * this Provider.
+     *
+     * @param loader The instance of [CsafLoader] used for fetching of online resources.
+     * @param channelCapacity The capacity of the channels used to buffer parallel fetches. Defaults
+     *   to [DEFAULT_CHANNEL_CAPACITY].
+     * @return The expected number of [RetrievedDocument]s provided.
+     */
+    @OptIn(ExperimentalCoroutinesApi::class)
+    suspend fun countExpectedDocuments(
```

Comment:
Is this actually used somewhere or just here for testing purposes? If its the latter, I would propose to move it to the test module

Reply:
It is meant to be actually used in the near future and therefore also has a Java counterpart.
Reason is that providers like RedHat hold enormous amounts of CSAF documents which take a considerable time to download. With that function, one can do a fast "preflight" to calculate approx. fetching progress of documents later on.

Reply:
> It is meant to be actually used in the near future and therefore also has a Java counterpart. Reason is that providers like RedHat hold enormous amounts of CSAF documents which take a considerable time to download. With that function, one can do a fast "preflight" to calculate approx. fetching progress of documents later on.

Got it, that makes sense